### DISCUSSION TITLE
Create a known IP address for WSL2 VM

### DISCUSSION THREAD

**Discussion by Biswa96** [microsoft/WSL#7395] [Show and tell]
⚠️ ⚠️ The following code is just for testing purposes. Please don't use this for any serious stuff. ⚠️ ⚠️ 

* The following code creates a Hyper-V virtual network with same name and same id used in WSL2.
* It is possible to add your own network subnet and probably many things. The subnet is `172.16.0.0/24` (for example).
* The program need to be executed **before** running any WSL2 instance. Otherwise, you have to delete the previous create virtual network.
* Use this well.

```cpp
#include <windows.h>
#include <computenetwork.h>

#pragma comment(lib, "computenetwork")

int main(void)
{
    HRESULT hRes;
    HCN_NETWORK hcnNetwork;
    PWSTR errRecord;
    const GUID netId = {
        0xB95D0C5E,
        0x57D4,
        0x412B,
        { 0xB5, 0x71, 0x18, 0xA8, 0x1A, 0x16, 0xE0, 0x05 } };

    // Flags = EnableDnsProxy + IsolateVSwitch
    // Type = Static
    PCWSTR netSettings = LR"(
    {
        "Name" : "WSL",
        "Type" : "ICS",
        "IsolateSwitch" : true,
        "Flags" : 9,
        "Subnets": [
            {
                "AddressPrefix":"172.16.0.0/24"
            }
        ]
    })";

    hRes = HcnCreateNetwork(netId, netSettings, &hcnNetwork, &errRecord);
    hRes = HcnCloseNetwork(hcnNetwork);

    return 0;
}
```

There are many things can be added or modify. Please make the code **free and open-source** if you modify it so that everyone can learn and share this knowledge.

**Comment by pieterhouwen** [comment]
Hi there!

Can this be used for https://github.com/microsoft/WSL/issues/7690 ?
Is there a way to create a new switch adapter and specify in the wsl2 distro which one to use?

**Reply by Biswa96** [reply]
No, there can only one virtual network adapter be attached to WSL2 VM. The `GUID netId` is unique. Bridging network interfaces may have inverse effect.

**Comment by skorhone** [comment]
@Biswa96 My (example) script reconfigures subnet that WSL2 uses using a static network range. This change doesn't have any major side-effects.

Unfortunately configuring static IP the right way would require configuring static MAC for WSL2 VM and static MAC to IP rule for DHCP server. All this probably is doable, but reverse-engineering isn't worth the time.

If you don't care about possible side-effects, you can set desired static IP by running a script inside the WSL2 vm. If you want to minimize risk of side-effects (IP conflict), prefer IP addresses from end of the network range

**Reply by Biswa96** [reply]
There is no reverse engineering. Evrything is documented by Microsoft. We've to just complete the puzzle. There will be no IP conflict because there is only one WSL2 VM for a specific PC. To set a specific IP address, it has to be configured first from Windows side. I shall share the sample code after some days.

**Reply by skorhone** [reply]
Microsoft documentation for HCN API is far from complete. I found it easier to extract required information from [microsoft/hcsshim](https://github.com/microsoft/hcsshim) than try to spell their cryptic documentation.

HCN under the hood uses ICS for assigning IP addresses. I wasn't able to find anything useful for HCN + ICS integration. If you do find related documentation, share it with others 🙂

I'm personally happy with the solution I made. I'm not using Wsl2 for replacing standard VMs. Vagrant + HyperV is much better suited for that. I use Wsl2 mainly as a platform to run tools (podman, python, bash... ) & this use case doesn't warrant static IPs.

My only issue with Wsl2 is that we're not able to control network range with a group policy. It'd probably take a day or two from microsoft to implement this ☹️

**Reply by Biswa96** [reply]
I mean this one https://docs.microsoft.com/en-us/windows-server/networking/technologies/hcn/hcn-top

**Comment by Biswa96** [comment]
⚠️ ⚠️ The following code is just for testing purposes. Please don't use this for any serious stuff. ⚠️ ⚠️ 

To add a specific IP address in WSL2 VM, follow this code.

* Remember to run this code after WSL2 VM starts.
* This code removes the existing virtual network adapter attached to WSL2 VM and attaches a new one with your provided configuration.
* Run the `ip` commands in WSL2 shown at the end of the code.

```cpp
#include <windows.h>
#include <computecore.h>
#include <computenetwork.h>

#ifdef _MSC_VER
#pragma comment(lib, "computecore")
#pragma comment(lib, "computenetwork")
#endif

static const GUID NetworkId = {
    0xB95D0C5E,
    0x57D4,
    0x412B,
    { 0xB5, 0x71, 0x18, 0xA8, 0x1A, 0x16, 0xE0, 0x05 } };

static const GUID EndpointId = {
    0x12345678,
    0x1234,
    0x1234,
    { 0x12, 0x34, 0x12, 0x34, 0x56, 0x78, 0x90, 0xAB } };

static const wchar_t NetworkSettings[] = LR"(
    {
        "Name" : "WSL",
        "Type" : "ICS",
        "IsolateSwitch" : true,
        "Flags" : 9,
        "Subnets": [
            {
                "AddressPrefix":"172.16.0.0/24"
            }
        ]
    })";

// VirtualNetwork should be same with NetworkId
// Others are random at your choice
static const wchar_t EndpointSettings[] = LR"(
    {
        "ID": "12345678-1234-1234-1234-1234567890AB",
        "IPAddress": "172.16.0.2",
        "MacAddress": "12-12-12-12-12-12",
        "VirtualNetwork" : "B95D0C5E-57D4-412B-B571-18A81A16E005",
        "Namespace" : {
            "ID" : "00000000-0000-0000-0000-000000000000",
            "IsDefault" : false
        }
    })";

static const wchar_t NicAdapterRemove[] = LR"(
    {
        "ResourcePath" : "VirtualMachine/Devices/NetworkAdapters/Default",
        "RequestType" : "Remove"
    })";

// EndpointId and MacAddress should match with EndpointSettings
static const wchar_t NicAdapterAdd[] = LR"(
    {
        "ResourcePath" : "VirtualMachine/Devices/NetworkAdapters/Default",
        "RequestType" : "Add",
        "Settings" : {
                "EndpointId" : "12345678-1234-1234-1234-1234567890AB",
                "MacAddress" : "12-12-12-12-12-12"
            }
    })";

int main(int argc, char *argv[])
{
    HRESULT hRes;
    HCN_NETWORK HcnNet;
    HCN_ENDPOINT HcnEndp;
    HCS_OPERATION HcsOp;
    HCS_SYSTEM HcsSys;
    PWSTR Result;

    // Put the VmId from `hcsdiag.exe list` command
    wchar_t VmId[] = L"";

    HcsOp = HcsCreateOperation(NULL, NULL);
    hRes = HcsOpenComputeSystem(VmId, GENERIC_ALL, &HcsSys);

    // 1. Delete existing network
    hRes = HcnDeleteNetwork(NetworkId, &Result);

    // 2. Create network and endpoint
    hRes = HcnCreateNetwork(NetworkId, NetworkSettings, &HcnNet, &Result);
    hRes = HcnCreateEndpoint(HcnNet, EndpointId, EndpointSettings, &HcnEndp, &Result);

    // 3. Detach existing network
    hRes = HcsModifyComputeSystem(HcsSys, HcsOp, NicAdapterRemove, NULL);
    hRes = HcsWaitForOperationResult(HcsOp, INFINITE, &Result);

    // 4. Add our own network
    hRes = HcsModifyComputeSystem(HcsSys, HcsOp, NicAdapterAdd, NULL);
    hRes = HcsWaitForOperationResult(HcsOp, INFINITE, &Result);

    hRes = HcnCloseEndpoint(HcnEndp);
    hRes = HcnCloseNetwork(HcnNet);

    HcsCloseOperation(HcsOp);
    HcsCloseComputeSystem(HcsSys);

    return 0;
}

// Now run these commands in WSL2:
// sudo ip address add 172.16.0.2/24 broadcast + dev eth0
// sudo ip link set dev eth0 up
// sudo ip route add default via 172.16.0.1 dev eth0
//
```

Feel free to modify and add this code in your project. Also ask any questions here. A proper attribution would be appreciated ❤️

**Comment by git-rz** [comment]
Hi @Biswa96 , I have tried using the code from the original post. It is not working for me. 

1) Installed Visual Studio 2022
2) Created a new c++ command line solution
3) Replaced the hello-world code with the code from above
4) Modified the address prefix to be 192.168.43.0/24
5) Reviewed all my startup programs and removed any that could possibly start wsl
6) Added the newly compiled command line tool to my startup folder
7) rebooted
8) logged in, waited for all startup programs to complete.
9) Opened WSL, ran `ip addr`, saw a 172 address.

I am using w10 enterprise, 21H2

Is there anything else that needs to be adapted to work on my computer?

**Reply by Biswa96** [reply]
The code in original post is just a proof-of-concept. Try not to use it in enterprise environment ⚠️ The return HRESULT values need to be checked. The program should be run as administrator and startup programs do not run as admin.

It is not required to reboot PC to modify network. 1. Shutdown any WSL instance if running. 2. Delete existing WSL2 network interface with this command in Powershell.

```
powershell.exe -NoProfile "Get-HnsNetwork | Where-Object {$_.Name -eq 'WSL'} | Remove-HnsNetwork"
```

I shall make a tool with all the modifications that can be done with WSL2 VM. Now I grasped a json parser in C++

**Comment by davidshen84** [comment]
Hi, could someone help me understand why we have to have a dynamic IP for the WSL instance? I've seen many posts talking about how to set the WSL IP to static. Why this cannot be part of the WSL official feature?

**Reply by Biswa96** [reply]
Only Microsoft developers can provide proper answer. But recently, there was a feature added to attach any virtual network switch. So, one can set the network range with that.

For my usecase, I have a small command line tool with the idea from original post.

**Reply by pieterhouwen** [reply]
Does this mean that you can assign multiple switches or does this just mean that you're not limited to the standard vEthernet switch?

Asking in regards to https://github.com/microsoft/WSL/discussions/7395#discussioncomment-1638449

**Reply by Biswa96** [reply]
I am not sure about that. The options are not officially documented. There are many users using them, see these issues https://github.com/microsoft/WSL/issues?q=is%3Aissue+is%3Aopen+vmswitch

**Comment by micheldiemer** [comment]
Here a detailed working procedure on how to use WSLAttachSwitch

Script : https://github.com/microsoft/WSL/issues/4799#issuecomment-2065450904

Text : https://github.com/microsoft/WSL/issues/4799#issuecomment-2065456777

**Comment by micheldiemer** [comment]
See also

- https://github.com/skorhone/wsl2-custom-network
- networkingMode=mirrored
- `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss\NatGatewayIpAddress Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss\NatNetwork Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss`
